### PR TITLE
Create Helm templates from resource manifests

### PR DESCRIPTION
**Pull request by itowlson**
This PR provides support for authoring Helm templates based on existing resources.  The intended workflow is as follows:

* Open a resource in the cluster explorer tree view and invoke the `Helm: Convert to Template` command.  (You can also right-click the resource and choose `Convert to Template` from the context menu.
* You are prompted for a template name (and, if necessary, a chart to add it to).
* A template is created in the chart's `templates` directory.  Initially, we only 'templatise' the `metadata.name` and `metadata.labels.chart` fields.
* You can now move the cursor to any value in the template and invoke the `Helm: Convert to Template Parameter` command.
* The extension replaces the value with a template expression (`{{ .Values.widget.snoofles }}`) and creates a corresponding entry and value (`widget: > snoofles: mysnooflevalue`) in values.yaml.
* You can continue converting values to parameters until your template is PERFECT IN EVERY WAY.

The scope for this is 'basic getting started'.  It's not intended to be a means of maintaining complex charts - just a way for someone to take an existing resource and say "this does what I want but I just need to be able to vary these knobs."

Now, I have plenty of questions about my implementation.

* Is this a useful workflow for Helm chart authors?
* Are there any values which should _always_ be parameterised in predictable ways?
* Is it okay to infer a parameter name or is it important to offer fine control?
* Is the structure of the generated parameters reasonable?  It's a bit flatter than the manifest YAML but that seems consistent with some of the sample templates.
* Are the resultant templates and `values.yaml` entries correct, or have I missed some subtlety of how Helm wires these things up?
* Does anything about the output make you flinch and go "ugh, we don't want people doing it like _that_"?  (The implementation code _will_ make you flinch but let's skate delicately past that.)

It would be great if folks could have a play with it and see if it feels natural and helpful.  Please also report bugs as it's not been extensively tested, though at this stage I'm most concerned about getting the right flow - squashing the edge cases can come later.  Things in the 'and it would be great if it could do X' category might be best raised as issues so that we don't try to boil the ocean in one PR - it's already large and complicated enough - but I can incorporate them if they're must-have.

### CODE REVIEW

**Comment by bacongobbler** [comment]
will take a spin when back on Tuesday for sure!

**Comment by bacongobbler** [comment]
I found one bug: if you convert a deployment to a template and give it a name with hypens (say, `brigade-api-deployment`), then when you convert a value in that template using `Helm: Convert to Template`, it'll use the template name as part of the value. That's normally fine, but hyphens are a special case in Go templates. There's a short write-up of it here: https://github.com/helm/helm/issues/2192

Since it's just interaction with the values, would it be best to convert invalid Go template characters to underscores?

e.g. if I have a template called `brigade-api-deployment` and convert a value to a template parameter, the resultant value should be `.Values.brigade_api_deployment.myValue`.

What do you think?

**Comment by bacongobbler** [comment]
This, by the way, is INSANELY useful, and users have been asking for a feature like this for a long time. See https://github.com/helm/helm/issues/2730 for a proposal on the same features, but as a Helm plugin.

cc @thomastaylor312, you'd probably get a kick out of this

**Comment by bacongobbler** [comment]
To answer your questions:

> Now, I have plenty of questions about my implementation.
> 
>     Is this a useful workflow for Helm chart authors?

Hard yes.

>     Are there any values which should always be parameterised in predictable ways?

Nope! Nothing is set in stone in the [Chart Template Guide](https://docs.helm.sh/chart_template_guide) either.

>     Is it okay to infer a parameter name or is it important to offer fine control?

I could see potential value in prompting while providing the default value, but I like the current infer model as it makes for a better UX for someone unfamiliar with the Helm chart spec. The convention VSCode is choosing looks fantastic to me, at least. If the user wants to customize the value, they can make changes to the template and value after the fact.

>     Is the structure of the generated parameters reasonable? It's a bit flatter than the manifest YAML but that seems consistent with some of the sample templates.

yep!

>     Are the resultant templates and values.yaml entries correct, or have I missed some subtlety of how Helm wires these things up?

See my mini bug report above. Otherwise it looks good to me!

>     Does anything about the output make you flinch and go "ugh, we don't want people doing it like that"? (The implementation code will make you flinch but let's skate delicately past that.)

Absolutely not. The experience is great.

**Comment by technosophos** [comment]
I also like the inferring of the name.

**Comment by itowlson** [comment]
Tidied some code and improved some error handling.  It's not perfect, but ready for a review I think.  @bnookala do you have time?  Thanks!

**Comment by thomastaylor312** [comment]
Oh wow, this is super nice. Especially because I haven't had any time to work on the commandeer stuff.

File: src/helm.symbolProvider.ts [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+export function symbolAt(position: vscode.Position, sis: vscode.SymbolInformation[]): vscode.SymbolInformation | undefined {
+    const containers = sis.filter((si) => si.location.range.contains(position));
+    if (containers.length === 0) {
+        return undefined;
+    }
+    return minimal(containers);
+}
+
+function minimal(sis: vscode.SymbolInformation[]): vscode.SymbolInformation {
```
Comment (bnookala) [inline src/helm.symbolProvider.ts]: can this be more descriptive? minimal what?

**Comment by bacongobbler** [comment]
One other edge case: when converting secrets to templates, the base64 encoding of the secret's data is retained.

Basically, when I convert to a template, this is what I get:

```
apiVersion: v1
data:
  allowHostMounts: IEhlbHAsIEknbSBiZWluZyBvcHByZXNzZWQuIENvbWUgYW5kIHNlZSB0aGUgdmlvbGVuY2UgaW5oZXJlbnQgaW4gdGhlIHN5c3RlbS4=
  allowPrivilegedJobs: IEhlbHAsIEknbSBiZWluZyBvcHByZXNzZWQuIENvbWUgYW5kIHNlZSB0aGUgdmlvbGVuY2UgaW5oZXJlbnQgaW4gdGhlIHN5c3RlbS4=
  cloneURL: IEhlbHAsIEknbSBiZWluZyBvcHByZXNzZWQuIENvbWUgYW5kIHNlZSB0aGUgdmlvbGVuY2UgaW5oZXJlbnQgaW4gdGhlIHN5c3RlbS4=
  initGitSubmodules: IEhlbHAsIEknbSBiZWluZyBvcHByZXNzZWQuIENvbWUgYW5kIHNlZSB0aGUgdmlvbGVuY2UgaW5oZXJlbnQgaW4gdGhlIHN5c3RlbS4=
  repository: IEhlbHAsIEknbSBiZWluZyBvcHByZXNzZWQuIENvbWUgYW5kIHNlZSB0aGUgdmlvbGVuY2UgaW5oZXJlbnQgaW4gdGhlIHN5c3RlbS4=
  sharedSecret: IEhlbHAsIEknbSBiZWluZyBvcHByZXNzZWQuIENvbWUgYW5kIHNlZSB0aGUgdmlvbGVuY2UgaW5oZXJlbnQgaW4gdGhlIHN5c3RlbS4=
  vcsSidecar: IEhlbHAsIEknbSBiZWluZyBvcHByZXNzZWQuIENvbWUgYW5kIHNlZSB0aGUgdmlvbGVuY2UgaW5oZXJlbnQgaW4gdGhlIHN5c3RlbS4=
kind: Secret
metadata:
  annotations:
    projectName: bacongobbler/test
  labels:
    app: brigade
    chart: "{{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}"
    component: project
    heritage: Tiller
    release: test
  name: {{ template "fullname" . }}
type: brigade.sh/project
```

This means that if I install this chart, the secret's metadata will get base64-encoded a second time, which probably isn't the user's intent.

**Comment by itowlson** [comment]
@bacongobbler What is the desired behaviour in the secrets scenario then?  Should all data values be automatically converted to parameters or something?

File: src/helm.authoring.ts [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+
+const NAME_EXPRESSION = '{{ template "fullname" . }}';
+const CHART_LABEL_EXPRESSION = '{{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}';
+
+const QUOTE_CONTROL_INFO = [
+    { text: NAME_EXPRESSION, mode: QuoteMode.None },
+    { text: CHART_LABEL_EXPRESSION, mode: QuoteMode.Double },
+];
+
+function foldSpindleAndMutilate(template: any): void {
```
[the remark this answers is not in the document]
Comment (itowlson) [reply src/helm.authoring.ts]: _embarrassment_  oh yeah I was gonna fix that

File: src/helm.authoring.ts [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+
+export interface TextEdit {
+    readonly document: vscode.TextDocument;
+    readonly edits: vscode.TextEdit[];
+}
+
+export async function convertToParameter(fs: FS, host: Host, document: vscode.TextDocument, selection: vscode.Selection): Promise<Errorable<TextEdit>> {
+    const helmSymbols = await getHelmSymbols(document);
+    if (helmSymbols.length === 0) {
+        return { succeeded: false, error: ['Active document is not a Helm template'] };
```
[the remark this answers is not in the document]
Comment (itowlson) [reply src/helm.authoring.ts]: Pull which into its own function?

**Comment by itowlson** [comment]
I can try base64-decoding the `data` fields but not sure what to do if the result is not text.  Or is that not a concern?  I also see incantations like `{{ default "" .Values.api.password | b64enc | quote }}` in public Helm charts - do I need to reproduce this and under what circumstances?

**Comment by technosophos** [comment]
I think it only gets re-encoded for the `stringData` key. The `data` key will simply leave it alone. Or am I missing something?

**Comment by itowlson** [comment]
I tried round-tripping a secret and it _seemed_ to work but I'm not sure what additional shenanigans Helm might be putting in its way.  It definitely looks like `kubectl create` (when passed a YAML of kind `Secret`) expects secrets data to be base64 encoded in the YAML.

I'm inclined to merge with this as it stands in order to get it into the release, and raise an issue for this once we can pin down the scope of the issue and desired behaviour.

**Comment by itowlson** [comment]
Okay, after discussions we believe that the existing code handles secrets data correctly, and I have a thumbs of from Bhargav to defer addressing his remaining comments, so I'm merging this quick before anyone turns up anything else!  #supremeprofessionalism
